mirgecom/filter.py

```diff
@@ -0,0 +1,194 @@
+    mode_response_function:
+        A function that returns a filter weight for for each mode id.
+
+    Returns
+    -------
+    filter: np.ndarray
+        filter operator in the modal basis
+    """
+    mode_ids = element_group.mode_ids()
+    order = element_group.order
+    dim = element_group.dim
+
+    nmodes = len(mode_ids)
+    filter = np.identity(nmodes)
+    nfilt = order - cutoff
+    if nfilt <= 0:
+        return filter
+
+    for mode_index, mode_id in enumerate(mode_ids):
+        mode = mode_id
+        if dim > 1:
+            mode = sum(mode_id)
+        if mode >= cutoff:
+            filter[mode_index, mode_index] = mode_response_function(mode,
+                                                                    cutoff=cutoff,
+                                                                    nfilt=nfilt)
+    return filter
+
+
+def linear_operator_kernel():
+    """Apply linear operator to all elements."""
+    from meshmode.array_context import make_loopy_program
+    knl = make_loopy_program(
+        """{[iel,idof,j]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in}""",
+        "result[iel,idof] = sum(j, mat[idof, j] * vec[iel, j])",
+        name="spectral_filter")
+    knl = lp.tag_array_axes(knl, "mat", "stride:auto,stride:auto")
+    return knl
+
+
+@obj_array_vectorized_n_args
+def apply_linear_operator(discr, operator, fields):
+    """Apply *operator* matrix to *fields*."""
+    actx = fields.array_context
+    result = discr.empty(actx, dtype=fields.entry_dtype)
+    for group in discr.groups:
+        actx.call_loopy(
+            linear_operator_kernel(),
+            mat=actx.from_numpy(operator),
+            result=result[group.index],
+            vec=fields[group.index])
+    return result
+
+
+def create_group_filter_operator(group, cutoff, response_func):
+    """Create spectral filter operator for *group*."""
+    filter_mat = make_spectral_filter(group, cutoff,
+                                      mode_response_function=response_func)
+    from modepy import vandermonde
+    vander = vandermonde(group.basis(), group.unit_nodes)
+    vanderm1 = np.linalg.inv(vander)
+    filter_operator = vander @ filter_mat @ vanderm1
+    return filter_operator
+
+
+@obj_array_vectorized_n_args
+def filter_modally(discrwb, dd, cutoff, mode_resp_func, field):
+    """Stand-alone procedural interface to spectral filtering.
+
+    For each element group in the discretization, and restriction,
+    This routine generates:
+    * a filter operator:
+        - *cutoff* filters only modes above this mode id
+        - *mode_resp_func* function returns a filter coefficient
+        for a given mode
+        - memoized into the discretization
+    * a kernel to apply the operator
+        - memoized into the array context
+    * a filtered solution wherein the filter is applied to *field*.
+
+    Parameters
+    ----------
+    discrwb: :class:`grudge.discrwb`
```

Comment:
I'm surprised this passed the reference checker---that identifier doesn't exist. Do you have the latest master merged?

Reply:
Yeah, it's merged.  I was sloppy here because it auto-filled it.  Is it better now?